File: trip-kit/src/main/java/com/skedgo/android/tripkit/RegionService.java
package com.skedgo.android.tripkit;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.skedgo.android.common.model.Location;
import com.skedgo.android.common.model.ModeInfo;
import com.skedgo.android.common.model.Region;
import com.skedgo.android.common.model.TransportMode;

import java.util.List;
import java.util.Map;

import rx.Observable;

public interface RegionService {
  Observable<List<Region>> getRegionsAsync();
  Observable<Region> getRegionByLocationAsync(@Nullable Location location);
  Observable<Location> getCitiesAsync();
  Observable<Location> getCitiesByNameAsync(@Nullable String name);
  Observable<Map<String, TransportMode>> getTransportModesAsync();
  Observable<List<TransportMode>> getTransportModesByIdsAsync(@NonNull List<String> modeIds);
  Observable<List<TransportMode>> getTransportModesByRegionAsync(@NonNull Region region);
  Observable<List<TransportMode>> getTransportModesByLocationAsync(@NonNull Location location);
  Observable<Void> refreshAsync();

  Observable<RegionInfo> getRegionInfoByRegionAsync(@NonNull Region region);
  Observable<Paratransit> fetchParatransitByRegionAsync(@NonNull Region region);
  Observable<List<ModeInfo>> getTransitModesByRegionAsync(@NonNull Region region);
}

File: TripKitAndroid/src/main/java/com/skedgo/android/tripkit/RegionService.kt
package com.skedgo.android.tripkit

import com.skedgo.android.common.model.Location
import com.skedgo.android.common.model.Region
import com.skedgo.android.common.model.TransportMode
import com.skedgo.android.tripkit.tsp.Paratransit
import com.skedgo.android.tripkit.tsp.RegionInfo

import rx.Observable
import skedgo.tripkit.routing.ModeInfo

interface RegionService {
  fun getRegionsAsync(): Observable<List<Region>>
  fun getRegionByNameAsync(regionName: String): Observable<Region>
  fun getRegionByLocationAsync(latitude: Double, longitude: Double): Observable<Region>
  fun getRegionByLocationAsync(location: Location?): Observable<Region>
  fun getCitiesAsync(): Observable<Location>
  fun getCitiesByNameAsync(name: String?): Observable<Location>
  fun getTransportModesAsync(): Observable<Map<String, TransportMode>>
  fun getTransportModeByIdAsync(modeId: String): Observable<TransportMode>
  fun getTransportModesByIdsAsync(modeIds: List<String>): Observable<List<TransportMode>>
  fun getTransportModesByRegionAsync(region: Region): Observable<List<TransportMode>>
  fun getTransportModesByLocationAsync(location: Location): Observable<List<TransportMode>>
  fun refreshAsync(): Observable<Void>
  fun getRegionInfoByRegionAsync(region: Region): Observable<RegionInfo>
  fun fetchParatransitByRegionAsync(region: Region): Observable<Paratransit>
  fun getTransitModesByRegionAsync(region: Region): Observable<List<ModeInfo>>
}